Incorrect above-the-line-deduction for qualified tuition expenses before 2021

Before 2021, qualified tuition expenses could be taken as an above-the-line-deduction lowering  AGI.  Turbo Tax believes the amount of tuition expenses that can be used to reduce AGI is limited to $4000 per person with tuition expenses as shown in this screenshot:

<img width="1742" alt="Screen Shot 2022-09-01 at 1 02 01 PM" src="https://user-images.githubusercontent.com/12170745/188014324-c5e61492-c7d8-4d22-bf4e-2ffafdbed5d4.png">

The following simple YAML test file shows that OpenFisca-US 0.129.2 version does not cap the above-the-line-deduction, and therefore, reduces AGI by too much.   Here is test case:
```
- name: Tax unit created by hand for AGI test
  absolute_error_margin: 0.01
  period: 2020
  input:
    people:
      person1:
        is_tax_unit_head: 1
        is_tax_unit_spouse: 0
        is_tax_unit_dependent: 0
        age: 25
        ssi: 0
        state_supplement: 0
        wic: 0
        employment_income: 74000
        qualified_tuition_expenses: 14000
        is_eligible_for_american_opportunity_credit: 1
    spm_units:
      spm_unit:
        members: [person1]
        snap: 0
        tanf: 0
    tax_units:
      tax_unit:
        members: [person1]
        premium_tax_credit: 0
    households:
      household:
        members: [person1]
        state_code: TX
  output:
    adjusted_gross_income: 70000
```

Here are the test results for the above case:
```
================================= FAILURES ===================================
_______________________________ test session _________________________________
/Users/mrh/work/m20-id0.test.yaml:
  Test 'Tax unit created by hand for AGI test':
    adjusted_gross_income@2020: [60000.] differs from 70000.0 
      with an absolute margin [10000.] > 0.01
========================= short test summary info ============================
FAILED m20-id0.test.yaml::
============================ 1 failed in 0.49s ===============================
```

And here are some excerpts from the test trace output:
```
irs_gross_income<2020> >> [74000.]
qualified_tuition_expenses<2020> >> [14000.]
above_the_line_deductions<2020> >> [14000.]
adjusted_gross_income<2020> >> [60000.]
```

The USI-Tax-Analyzer generates an AGI amount of $70,000 for this case.  This case cannot be simulated by TAXSIM35 because that model does not have qualified tuition expenses as an input variable.

--- a/changelog.d/fix-issue-1318.fixed.md
+++ b/changelog.d/fix-issue-1318.fixed.md
@@ -1,0 +1,1 @@
+Applied the IRS Form 8917 cap to the pre-2021 tuition above-the-line deduction, which previously flowed into AGI uncapped.

--- a/policyengine_us/parameters/gov/irs/ald/deductions.yaml
+++ b/policyengine_us/parameters/gov/irs/ald/deductions.yaml
@@ -7,7 +7,7 @@ values:
     - early_withdrawal_penalty
     - alimony_expense
     - educator_expense
-    - qualified_tuition_expenses
+    - capped_qualified_tuition_expenses_ald
     - domestic_production_ald
     - health_savings_account_ald
     - self_employed_health_insurance_ald
@@ -24,7 +24,7 @@ values:
     - early_withdrawal_penalty
     - alimony_expense
     - educator_expense
-    - qualified_tuition_expenses
+    - capped_qualified_tuition_expenses_ald
     - health_savings_account_ald
     - self_employed_health_insurance_ald
     - self_employed_pension_contribution_ald

--- a/policyengine_us/tests/policy/baseline/gov/irs/income/taxable_income/adjusted_gross_income/above_the_line_deductions/capped_qualified_tuition_expenses_ald.yaml
+++ b/policyengine_us/tests/policy/baseline/gov/irs/income/taxable_income/adjusted_gross_income/above_the_line_deductions/capped_qualified_tuition_expenses_ald.yaml
@@ -1,0 +1,99 @@
+# Regression tests for issue #1318.
+# Before 2021, the above-the-line deduction for qualified tuition expenses
+# (IRS Form 8917) was capped at $4,000 (or $2,000 for higher-MAGI filers).
+# The raw `qualified_tuition_expenses` variable previously flowed straight
+# into `above_the_line_deductions`, bypassing the cap. After the fix, the
+# contribution goes through `capped_qualified_tuition_expenses_ald`, which
+# applies the Form 8917 cap using a locally computed MAGI.
+
+- name: Capped tuition ALD at low income hits the $4,000 Form 8917 cap (issue #1318)
+  period: 2020
+  input:
+    people:
+      person1:
+        is_tax_unit_head: true
+        age: 25
+        employment_income: 44_000
+        qualified_tuition_expenses: 14_000
+        is_eligible_for_american_opportunity_credit: false
+    tax_units:
+      tax_unit:
+        members: [person1]
+        premium_tax_credit: 0
+    households:
+      household:
+        members: [person1]
+        state_code: TX
+  output:
+    # MAGI for Form 8917 is $44,000 (gross income with no other ALDs);
+    # that is below the $65,000 single threshold, so the cap is $4,000.
+    # Before the fix, ALD pulled in the full $14,000 raw amount.
+    capped_qualified_tuition_expenses_ald: 4_000
+    above_the_line_deductions: 4_000
+    adjusted_gross_income: 40_000
+
+- name: Capped tuition ALD is zero above the $80k MAGI phase-out for a 2020 single filer (issue #1318)
+  period: 2020
+  input:
+    people:
+      person1:
+        is_tax_unit_head: true
+        age: 25
+        employment_income: 100_000
+        qualified_tuition_expenses: 5_000
+        is_eligible_for_american_opportunity_credit: false
+    tax_units:
+      tax_unit:
+        members: [person1]
+        premium_tax_credit: 0
+    households:
+      household:
+        members: [person1]
+        state_code: TX
+  output:
+    capped_qualified_tuition_expenses_ald: 0
+    above_the_line_deductions: 0
+
+- name: Capped tuition ALD is zero starting 2021 after the statute was repealed (issue #1318)
+  period: 2021
+  input:
+    people:
+      person1:
+        is_tax_unit_head: true
+        age: 25
+        employment_income: 44_000
+        qualified_tuition_expenses: 14_000
+        is_eligible_for_american_opportunity_credit: false
+    tax_units:
+      tax_unit:
+        members: [person1]
+        premium_tax_credit: 0
+    households:
+      household:
+        members: [person1]
+        state_code: TX
+  output:
+    capped_qualified_tuition_expenses_ald: 0
+    above_the_line_deductions: 0
+
+- name: Capped tuition ALD is zero for married filing separately (issue #1318)
+  period: 2020
+  input:
+    people:
+      person1:
+        is_tax_unit_head: true
+        age: 25
+        employment_income: 44_000
+        qualified_tuition_expenses: 14_000
+        is_eligible_for_american_opportunity_credit: false
+    tax_units:
+      tax_unit:
+        members: [person1]
+        filing_status: SEPARATE
+        premium_tax_credit: 0
+    households:
+      household:
+        members: [person1]
+        state_code: TX
+  output:
+    capped_qualified_tuition_expenses_ald: 0

--- a/policyengine_us/variables/gov/irs/income/taxable_income/adjusted_gross_income/above_the_line_deductions/capped_qualified_tuition_expenses_ald.py
+++ b/policyengine_us/variables/gov/irs/income/taxable_income/adjusted_gross_income/above_the_line_deductions/capped_qualified_tuition_expenses_ald.py
@@ -1,0 +1,55 @@
+from policyengine_us.model_api import *
+
+
+class capped_qualified_tuition_expenses_ald(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Qualified tuition expenses above-the-line deduction (capped)"
+    unit = USD
+    definition_period = YEAR
+    documentation = (
+        "Tuition and fees above-the-line deduction from IRS Form 8917 "
+        "(pre-2021 only). Applies the IRC 222(b)(2) cap using a locally "
+        "computed MAGI to avoid a circular dependency on "
+        "`adjusted_gross_income`."
+    )
+    reference = (
+        "https://www.irs.gov/pub/irs-pdf/f8917.pdf#page=2"
+        # Law was repealed starting tax year 2021.
+        "https://irc.bloombergtax.com/public/uscode/doc/irc/section_222"
+    )
+
+    def formula(tax_unit, period, parameters):
+        # This variable contributes to the ALD list only in years when
+        # `capped_qualified_tuition_expenses_ald` is listed in
+        # `gov.irs.ald.deductions`. If a reform removes it, return 0.
+        all_alds = parameters(period).gov.irs.ald.deductions
+        if "capped_qualified_tuition_expenses_ald" not in all_alds:
+            return 0
+        # Married filing separately cannot take the tuition deduction.
+        filing_status = tax_unit("filing_status", period)
+        separate = filing_status == filing_status.possible_values.SEPARATE
+        qualified_tuition_expenses = add(
+            tax_unit, period, ["qualified_tuition_expenses"]
+        )
+        # Build a simple MAGI from non-dependent person-level income sources,
+        # excluding sources that depend on MAGI themselves (taxable SS, UI).
+        irs = parameters(period).gov.irs
+        gross_income_sources = irs.gross_income.sources
+        person = tax_unit.members
+        not_dependent = ~person("is_tax_unit_dependent", period)
+        cycle_sources = {
+            "taxable_social_security",
+            "taxable_unemployment_compensation",
+        }
+        safe_sources = [src for src in gross_income_sources if src not in cycle_sources]
+        if "taxable_unemployment_compensation" in gross_income_sources:
+            safe_sources.append("unemployment_compensation")
+        magi = 0
+        for source in safe_sources:
+            magi += not_dependent * max_(0, add(person, period, [source]))
+        magi = tax_unit.sum(magi)
+        p = parameters(period).gov.irs.deductions.tuition_and_fees
+        joint = filing_status == filing_status.possible_values.JOINT
+        cap = where(joint, p.joint.calc(magi), p.non_joint.calc(magi))
+        return where(separate, 0, min_(qualified_tuition_expenses, cap))

